pkg/virt-handler/rest/console.go

```diff
@@ -45,17 +46,75 @@ type ConsoleHandler struct {
 	serialLock           *sync.Mutex
 	vncLock              *sync.Mutex
 	vmiInformer          cache.SharedIndexInformer
+	usbredir             [v1.UsbredirMaxNumberOfDevices]UsbredirHandler
+}
+
+type UsbredirHandler struct {
+	lock      *sync.Mutex
+	stopChans map[types.UID](chan struct{})
+	inUse     bool
 }
 
 func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiInformer cache.SharedIndexInformer) *ConsoleHandler {
-	return &ConsoleHandler{
+	c := &ConsoleHandler{
 		podIsolationDetector: podIsolationDetector,
 		serialStopChans:      make(map[types.UID](chan struct{})),
 		vncStopChans:         make(map[types.UID](chan struct{})),
 		serialLock:           &sync.Mutex{},
 		vncLock:              &sync.Mutex{},
 		vmiInformer:          vmiInformer,
 	}
+
+	for i := 0; i < v1.UsbredirMaxNumberOfDevices; i++ {
+		c.usbredir[i] = UsbredirHandler{
+			lock:      &sync.Mutex{},
+			stopChans: make(map[types.UID](chan struct{})),
+			inUse:     false,
+		}
+	}
+
+	return c
+}
+
+func (t *ConsoleHandler) USBRedirHandler(request *restful.Request, response *restful.Response) {
+	vmi, code, err := getVMI(request, t.vmiInformer)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Failed to retrieve VMI")
+		response.WriteError(code, err)
+		return
+	}
+
+	// Get the next free usbredir socket
+	slot := -1
+	for i, handler := range t.usbredir {
+		if !handler.inUse {
+			slot = i
+			break
+		}
+	}
+	if slot < 0 {
+		log.Log.Object(vmi).Reason(err).Error("All USB Redir devices are being used")
+		response.WriteError(http.StatusBadRequest, err)
```

Comment:
Maybe `StatusTooManyRequests` is more appropriate to this case? I'm not sure.. what do you think?

Reply:
Makes sense to me, altough when I tested e.g: a 3rd device when there is only 2 usbredir slots, I did not get the actual error from neither StatusBadRequest or StatusTooManyRequests from ConsoleHandler.stream(), perhaps some other fixes are needed to propagate that error to the client?